src/batMap.ml

```diff
@@ -806,6 +892,38 @@ module Concrete = struct
         | Some v2 -> add k (f v1 v2) cmp1 m)
         m1 empty
 
+  let add_seq cmp s m =
+    BatSeq.fold_left
+      (fun m (k, v) -> add k v cmp m)
+      m
+      s
+    
+  let of_seq cmp s =
+    add_seq cmp s empty
+
+  let rec seq_of_iter m () =
+    match m with
+    | E -> BatSeq.Nil
+    | C(k, v, r, e) ->
+       BatSeq.Cons ((k, v), seq_of_iter (cons_iter r e))
+      
+  let to_seq m =
+    seq_of_iter (cons_iter m E)
+      
+  let to_seq_from cmp k m = 
+    seq_of_iter (cons_iter_from cmp k m E)
+
+  let union_stdlib cmp f m1 m2 =
+    foldi
+      (fun k v m ->
+        update_stdlib k
+          (fun v2opt ->
+            match v2opt with
+            | Some v2 -> f k v v2
+            | None -> Some v) cmp m)
+      m1
+      m2
```

Comment:
This implementation does not have the expected complexity. I think that you could just use `merge` instead, which is more general. (A hand-written implementation of `union` using the same approach as `merge` would also be faster.)

Reply:
i used heuristic_merge instead. That also avoids problems when PMaps with incompatible comparisons are merged (see your comment below)